File: CODE/T4014/T4014.cpp
#define nametask "T4014"
#include <bits/stdc++.h>
#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define SZ(a) (int) a.size()
#define all(a) a.begin(), a.end()
#define FOR(i, a, b) for (int i = a; i <= b; i++)
#define FOD(i, b, a) for (int i = b; i >= a; i--)

using namespace std;
typedef long long ll;
typedef pair <int, int> pi;

ll f[20][172];
ll res[22], A[22], B[22];

vector <int> getdigit(ll x){
    vector <int> res;
    while (x > 0){
        res.pb(x % 10);
        x /= 10;
    }
    return res;
}

ll calc(int id, bool tight, int sum, int d, vector <int> digit){
    if (id < 0) return (sum == d);
    if (!tight && f[id][sum] != -1) return f[id][sum];
    int k = (tight) ? digit[id] : 9;
    ll cnt = 0;
    FOR(i, 0, k){
        bool new_tight = (i == digit[id]) ? tight : 0;
        cnt += calc(id - 1, new_tight, sum + i, d, digit);
    }
    if (tight) return cnt;
    return (f[id][sum] = cnt);
}

ll solve(ll a, ll b, int d){
    vector <int> digitA, digitB;
    digitA = getdigit(a - 1);
    digitB = getdigit(b);
    return calc(SZ(digitB) - 1, 1, 0, d, digitB) - calc(SZ(digitA) - 1, 1, 0, d, digitA);
}

signed main()
{   
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    if (fopen(nametask".inp", "r")){
        freopen(nametask".inp", "r", stdin);
        freopen(nametask".out", "w", stdout);
    }
    int t;
    cin >> t;
    FOR(i, 1, t){
        cin >> A[i] >> B[i];
    }
    vector <int> digitM = getdigit((ll)1e19);
    FOR(i, 1, 9 * 19){
        memset(f, -1, sizeof f);
        calc(SZ(digitM) - 1, 1, 0, i, digitM);
        FOR(j, 1, t) res[j] += solve((A[j] + i - 1) / i, B[j] / i, i);
    }
    FOR(i, 1, t) cout << res[i] << '\n';
    return 0;
}

File: CODE/T4014/T4014.py
import sys

sys.stdin = open("T4014.inp", "r")
sys.stdout = open("T4014.out", "w")

def get_digits(x):
    digits = []
    while x > 0:
        digits.append(x % 10)
        x //= 10
    return digits

def calc(idx, tight, digit_sum, d, digits, dp):
    if idx < 0:
        return 1 if digit_sum == d else 0
    
    if not tight and dp[idx][digit_sum] != -1:
        return dp[idx][digit_sum]
    
    limit = digits[idx] if tight else 9
    count = 0
    
    for i in range(limit + 1):
        new_tight = tight and (i == digits[idx])
        count += calc(idx - 1, new_tight, digit_sum + i, d, digits, dp)
    
    if not tight:
        dp[idx][digit_sum] = count
    
    return count

def solve(a, b, d, dp):
    digits_a = get_digits(a - 1)
    digits_b = get_digits(b)
    return calc(len(digits_b) - 1, True, 0, d, digits_b, dp) - calc(len(digits_a) - 1, True, 0, d, digits_a, dp)


t = int(input())
queries = [tuple(map(int, input().split())) for _ in range(t)]

max_digit_sum = 9 * 19
max_digit_length = 19
dp = [[-1] * (max_digit_sum + 1) for _ in range(max_digit_length)]

results = [0] * (t + 1)
digitM = get_digits(10**19)

for i in range(1, max_digit_sum + 1):
    dp = [[-1] * (max_digit_sum + 1) for _ in range(max_digit_length)]  # Reset memory
    calc(len(digitM) - 1, True, 0, i, digitM, dp)
    for j in range(t):
        a, b = queries[j]
        results[j] += solve((a + i - 1) // i, b // i, i, dp)

for res in results[:-1]:
    print(res)
